README.md
# **!!! Important Message !!!**

This project is being relicensed from a combination of the
"GNU LESSER GENERAL PUBLIC LICENSE 3.0" and "Mozilla Public License 2.0"
to the Bouncy Castle license. The new license can be read in the LICENSE file.

Subsequent to this change, this library will be transitioned into the
Bouncy Castle project. The Bouncy Castle team will take over the maintenance
and development of this library.

See https://github.com/bcgit/pc-dart for the new code.


Pointy Castle
=============

A Dart library for encryption and decryption. In this release, most of the classes
are ports of Bouncy Castle from Java to Dart. The porting is almost always
direct except for some classes that had been added to ease the use of low level
data.

To make sure nothing fails, tests and benchmarks for every algorithm are
provided. The expected results are taken from the Bouncy Castle Java version
and also from standards, and matched against the results got from Pointy Castle.

## Algorithms

In this release, the following algorithms are implemented:

**Block ciphers:**
  * AES

**Asymmetric block ciphers:**
  * RSA

**Asymmetric block cipher encodings:**
  * PKCS1
  * OAEP

**Stream ciphers:**
  * Salsa20

**Block cipher modes of operation:**
  * CBC (Cipher Block Chaining mode)
  * CFB (Cipher Feedback mode)
  * ECB (Electronic Code Book mode)
  * GCTR (GOST 28147 OFB counter mode)
  * OFB (Output FeedBack mode)
  * CTR (Counter mode)
  * SIC

**Paddings:**
  * PKCS7
  * ISO7816-4

**Digests:**
  * Blake2b
  * MD2
  * MD4
  * MD5
  * RIPEMD-128|160|256|320
  * SHA-1
  * SHA-224|256|384|512
  * SHA-512/t (t=8 to 376 and 392 to 504 in multiples of 8)
  * Keccak-224|256|384|512*
  * Tiger
  * Whirlpool

*_Keccak is currently implemented as SHA3Digest._

**MACs:**
  * HMAC
  * CMAC

**Signatures:**
  * (DET-)ECDSA
  * RSA

**Password based key derivators:**
  * PBKDF2
  * scrypt

**Asymmetric key generators:**
  * ECDSA
  * RSA

**Secure PRNGs:**
  * Based on block cipher in CTR mode
  * Based on block cipher in CTR mode with auto reseed (for forward security)
  * Based on Fortuna algorithm

### Instantiating implementation objects

There are two ways to instantiate objects that implement the
algorithms:

- using the registry, or
- without the registry.

#### Using the registry

Using the registry, the algorithm name is provided to high-level class
factories.

This is especially convenient when an algorithm involves multiple
algorithm implementation classes to implement. All the necessary
classes can all be instantiated with a single name
(e.g. "HMAC/SHA-256" or "SHA-1/HMAC/PBKDF2"), and they are
automatically combined together with the correct values.

To use the registry, either import `pointycastle.dart` or
`export.dart`.  For example,

```dart
import "package:pointycastleold/pointycastle.dart";

final sha256 = Digest("SHA-256");
final sha1 = Digest("SHA-1");
final md5 = Digest("MD5");

final hmacSha256 = Mac("SHA-256/HMAC");
final hmacSha1 = Mac("SHA-1/HMAC");
final hmacMd5 = Mac("MD5/HMAC");

final derivator = KeyDerivator("SHA-1/HMAC/PBKDF2");

final signer = Signer("SHA-256/RSA");
```

#### Without the registry

Without the registry, each implementation class must be instantiated
using its constructor.

If an algorithm involves multiple algorithm implementation classes,
they each have to be individually instantiated and combined together
with the correct values.

To use the constructors, import `export.dart`.  For example,

``` dart
import "package:pointycastleold/export.dart";

final sha256 = SHA256Digest();
final sha1 = SHA1Digest();
final md5 = MD5Digest();

final hmacSha256 = HMac(SHA256Digest(), 64);
final hmacSha512 = HMac(SHA512Digest(), 128);
final hmacMd5 = HMac(MD5Digest(), 64);

final derivator = PBKDF2KeyDerivator(HMac(SHA256Digest(), 64));

final signer = RSASigner(SHA256Digest(), '0609608648016503040201');
```

#### Registry vs without registry

Using the registry means that all algorithms will be imported by
default, which can increase the compiled size of your program.

To avoid this, instantiate **all** classes directly by using their
constructors.

### Importing libraries

There are two main approaches for importing Point Castle libraries:

- only import _pointycastle.dart_ (which includes the high-level API and
  the interfaces); or
- only import _export.dart_ (which includes the high-level API, interfaces
  and all the implementation classes).
  
It is also possible to import _api.dart_ (the high-level API) and
selectively import individual implementation classes as they are
needed. But this method requires a lot more programmer effort, and no
longer has any advantage over simply importing _export.dart_ and
relying Dart's tree shaking to leave out code that is not needed.

The registry can be used with any of these approaches. The different
approaches only affects access to the implementation classes.

Therefore, when avoiding the code size overheads of the registry, be
careful not to accidentally use the registry. Since the single use of
a factory from the registry will cause the entire registry to be
loaded.

## Tutorials

Some articles on how to use some of Pointy Castle's features can be
found under the _tutorials_ directory in the sources.

- [Calculating a digest](https://github.com/pointycastleold/pointycastleold/blob/master/tutorials/digest.md) - calculating a hash or digest (e.g. SHA-256, SHA-1, MD5)
- [Calculating a HMAC](https://github.com/pointycastleold/pointycastleold/blob/master/tutorials/hmac.md) - calculating a hash-based message authentication code (e.g. HMAC-SHA256, HMAC-SHA1)
- [Using AES-CBC](https://github.com/pointycastleold/pointycastleold/blob/master/tutorials/aes-cbc.md) - block encryption and decryption with AES-CBC
- [Using RSA](https://github.com/pointycastleold/pointycastleold/blob/master/tutorials/rsa.md) - key generation, signing/verifying, and encryption/decryption
- Some [tips](https://github.com/pointycastleold/pointycastleold/blob/master/tutorials/tips.md) on using Pointy Castle

_Note: the above links are to the most recent versions on the master
branch on GitHub. They may be different from the version here._


CHANGELOG.md
Changelog
=========

#### Version TBD

* Fixed OAEPEncoding and PKCS1Encoding to use provided output offset value.
* Fixed RSA block length and offset checks in RSAEngine.processBlock.
* Fixed RSASigner.verifySignature to return false when signature is bad.

#### Version 1.0.2 (2019-11-15)

* Add non-Keccak SHA3 support
* Add CMAC support ("AES/CMAC")
* Add ISO7816-4 padding support
* Fixes in CBCBlockCipherMac and CMac

#### Version 1.0.1 (2019-02-20)

* Add Blake2b support

#### Version 1.0.0 (2018-12-17) (Dart SDK version 2.0)

* Support Dart 2 and Strong Mode
* Migrate from `package:bignum.BigInteger` to `dart:core.BigInt`
* Remove Quiver and fixnum dependency
* OAEP encoding for block ciphers


#### Version 0.10.0 (2016-02-04) (Dart SDK version 0.14.0)

* First Pointy Castle release.

* Reorganised file structure.

* Completely new Registry implementation that dynamically loads imported implementations using reflection.
  It is explained in [this commit](https://github.com/pointycastleold/pointycastleold/commit/2da75e5a8d7bdbf95d08329add9f13b9070b75d4).

* Migrated from unittest to test package.


### cipher releases

#### Version 0.8.0 (2014-??-??) (Dart SDK version ???)

* **[bug 80]** PaddedBlockCipher doesn't add padding when data length is a multiple of the block 
                size. This fix introduces a **BREAKING CHANGE** in PaddedBlockCipher specification.
                Read its API documentation to know about the changes.


#### Version 0.7.0 (2014-03-22) (Dart SDK version 1.3.0-dev.5.2)

* **[enh 15]** Implement stream cipher benchmarks.
* **[enh 64]** Benchmark and optimize digests.
* **[enh 74]** Make SHA-3 usable in terms of speed.

* **[bug 67]** Removed some unused code.
* **[bug 68]** Fix process() method of PaddedBlockCipher.
* **[bug 75]** Remove a registry dependency in the Scrypt algorithm.


lib/macs/hmac.dart
// Copyright (c) 2013-present, the authors of the Pointy Castle project
// This library is dually licensed under LGPL 3 and MPL 2.0.
// See file LICENSE for more information.

library pointycastleold.impl.mac.hmac;

import "dart:typed_data";

import "package:pointycastleold/api.dart";
import "package:pointycastleold/src/registry/registry.dart";
import "package:pointycastleold/src/impl/base_mac.dart";

/**
 * HMAC implementation based on RFC2104
 *
 * H(K XOR opad, H(K XOR ipad, text))
 */
class HMac extends BaseMac {
  static final FactoryConfig FACTORY_CONFIG =
      new DynamicFactoryConfig.suffix(Mac, "/HMAC", (_, Match match) {
    final String digestName = match.group(1);
    final int blockLength = _DIGEST_BLOCK_LENGTH[digestName];
    if (blockLength == null) {
      throw new RegistryFactoryException("Digest $digestName unknown for "
          "HMAC construction.");
    }
    return () {
      Digest digest = new Digest(digestName);
      return new HMac(digest, blockLength);
    };
  });

  //TODO make this more generic
  static final Map<String, int> _DIGEST_BLOCK_LENGTH = {
    "GOST3411": 32,
    "MD2": 16,
    "MD4": 64,
    "MD5": 64,
    "RIPEMD-128": 64,
    "RIPEMD-160": 64,
    "SHA-1": 64,
    "SHA-224": 64,
    "SHA-256": 64,
    "SHA-384": 128,
    "SHA-512": 128,
    "Tiger": 64,
    "Whirlpool": 64,
  };

  //TODO reindent

  static final _IPAD = 0x36;
  static final _OPAD = 0x5C;

  Digest _digest;
  int _digestSize;
  int _blockLength;

  Uint8List _inputPad;
  Uint8List _outputBuf;

  HMac(this._digest, this._blockLength) {
    _digestSize = _digest.digestSize;
    _inputPad = new Uint8List(_blockLength);
    _outputBuf = new Uint8List(_blockLength + _digestSize);
  }

  String get algorithmName => "${_digest.algorithmName}/HMAC";

  int get macSize => _digestSize;

  void reset() {
    // reset the underlying digest.
    _digest.reset();

    // reinitialize the digest.
    _digest.update(_inputPad, 0, _inputPad.length);
  }

  void init(covariant KeyParameter params) {
    _digest.reset();

    var key = params.key;
    var keyLength = key.length;

    if (keyLength > _blockLength) {
      _digest.update(key, 0, keyLength);
      _digest.doFinal(_inputPad, 0);

      keyLength = _digestSize;
    } else {
      _inputPad.setRange(0, keyLength, key);
    }

    _inputPad.fillRange(keyLength, _inputPad.length, 0);

    _outputBuf.setRange(0, _blockLength, _inputPad);

    _xorPad(_inputPad, _blockLength, _IPAD);
    _xorPad(_outputBuf, _blockLength, _OPAD);

    _digest.update(_inputPad, 0, _inputPad.length);
  }

  void updateByte(int inp) {
    _digest.updateByte(inp);
  }

  void update(Uint8List inp, int inpOff, int len) {
    _digest.update(inp, inpOff, len);
  }

  int doFinal(Uint8List out, int outOff) {
    _digest.doFinal(_outputBuf, _blockLength);
    _digest.update(_outputBuf, 0, _outputBuf.length);

    var len = _digest.doFinal(out, outOff);
    _outputBuf.fillRange(_blockLength, _outputBuf.length, 0);
    _digest.update(_inputPad, 0, _inputPad.length);

    return len;
  }

  void _xorPad(Uint8List pad, int len, int n) {
    for (var i = 0; i < len; i++) {
      pad[i] ^= n;
    }
  }
}
